### PR TITLE
Bulk create attribute value

### PR DESCRIPTION
**Pull request by 8r2y5**
I want to merge this change because it fixes https://github.com/saleor/saleor/issues/13700

Measurements were done by performing mutation `productBulkCreate` and logging db queries (counting them).
| SELECT | 1 product | 100 products |  optimized 1 products | optimized 100 products |
| --- | --- | --- | --- | --- | 
| account_user | 3 |  3 | 3 |  3 |
| channel_channel | 2 | 2 | 2 |  2 |
| product_producttype | 12 | 411 | 12 |  286 |
| product_category | 3 | 85 | 3 |  85 |
| product_product | 5 | 168 | 5 |  168 |
| warehouse_stock | 1 | 1 | 1 |  1 |
| attribute_attributeproduct | 4 | 52 | 4 | 100 |
| product_productvariant | 4 | 131 | 4 |  131 |
| attribute_attributevalue | 20 | 670 | 20 |  318 |
| warehouse_warehouse | 2 | 2 | 2 |  2 |
| attribute_attribute | 11 | 366 | 11 |  366 |
| attribute_attributevariant | 4 | 144 | 4 |  144 |
| attribute_assignedproductattribute | 4 | 60 | 4 |  33 |
| attribute_assignedproductattributevalue | 4 | 84 | 4 |  9 |
| attribute_assignedvariantattribute | 3 | 152 | 3 |  125 |
| attribute_assignedvariantattributevalue | 6 | 268 | 6 |  9 |

| UPDATE | 1 product | 100 products | optimized 1 products | optimized 100 products |
| --- | --- | --- | --- | --- |
| account_user | 1 | 1 | 1 | 1 |
| attribute_assignedproductattributevalue | 1 | 33 | 1 | 33 |
| attribute_assignedvariantattributevalue | 3 | 133 | 3 | 125 |

| INNER JOIN | 1 product | 100 products | optimized 1 product | optimized 100 products |
| --- | --- | --- | --- | --- |
| warehouse_warehouse | 1 | 1 | 1 | 1 |
| warehouse_channelwarehouse | 1 | 1 | 1 | 1 |
| channel_channel | 1 | 1 | 1 | 1 |
| attribute_attributevariant | 13 | 506 | 13 | 506 |
| attribute_attributeproduct | 3 | 77 | 3 | 77 | 
| attribute_assignedproductattributevalue | 3 | 51 | - | - |
| attribute_attributevalue | 6 | 194 | - | - | 
| attribute_assignedvariantattributevalue | 3 | 143 | - | - |

| OUTER JOIN | 1 product | 100 products | optimized 1 products | optimized 100 products |
| --- | --- | --- | --- | --- |
| warehouse_warehouse_shipping_zones | 1 | 1 | 1 | 1 |
| shipping_shippingzone | 1 | 1 | 1 | 1 |
| shipping_shippingzone_channels | 1 | 1 | 1 | 1 |
| channel_channel | 1 | 1 | 1 | 1 |
| warehouse_allocation | 1 | 1 | 1 | 1 |

| INSERT | 1 product | 100 products | optimized 1 products | optimized 100 products |
| --- | --- | --- | --- | --- | 
| product_product | 1 | 1 | 1 | 1 |
| attribute_attributevalue | 4 | 117 | 3 | 81 | 
| attribute_assignedproductattribute | 3 | 51 | 3 | 33 |
| attribute_assignedproductattributevalue | 1 | 33 | 1 | 33 |
| product_productvariant | 1 | 1 | 1 | 1 |
| attribute_assignedvariantattribute | 3 | 143 | 3 | 125 |
| attribute_assignedvariantattributevalue | 3 | 125 | 3 | 125 |
| warehouse_stock | 1 | 1 | 1 | 1 |

# Impact

- [ ] New migrations
- [ ] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [ ] Link to documentation:

# Pull Request Checklist

- [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
- [ ] Database queries are optimized and the number of queries is constant
- [ ] Database migrations are either absent or optimized for zero downtime
- [ ] The changes are covered by test cases
- [ ] All new fields/inputs/mutations have proper labels added (`ADDED_IN_X`, `PREVIEW_FEATURE`, etc.)
- [ ] All migrations have proper dependencies
- [ ] All indexes are added concurrently in migrations
- [ ] All RunSql and RunPython migrations have revert option defined

### CODE REVIEW

File: saleor/attribute/models/base.py [hunk 1]
```diff
@@ -215,6 +216,76 @@ def get_translated_keys(self):
...
+            defaults = obj.pop("defaults")
+            query_params.append(models.Q(**obj))
+            obj["defaults"] = defaults
+
+        query = self.filter(reduce(lambda x, y: x | y, query_params))
+
+        for obj in objects_data:
+            defaults = obj.pop("defaults")
+            try:
+                record = query.get(**obj)
```
Comment (tomaszszymanski129) [inline saleor/attribute/models/base.py:235]: avoid using `get()` because of SQL `LIMIT 21;`, use `filter().first()` instead.

File: saleor/attribute/models/base.py [hunk 2]
```diff
@@ -215,6 +216,76 @@ def get_translated_keys(self):
...
```
Comment (tomaszszymanski129) [inline saleor/attribute/models/base.py:265]: same

File: saleor/graphql/attribute/utils.py [hunk 3]
```diff
@@ -491,18 +509,27 @@ def _pre_save_swatch_value(
...
+                    None,
+                    value,
+                ),
+            )
 
         if id := attr_values.swatch.id:
             _, attr_value_id = from_global_id_or_error(id)
-            model = attribute_models.AttributeValue.objects.get(pk=attr_value_id)
-            if not model:
+            value = attribute_models.AttributeValue.objects.get(pk=attr_value_id)
```
Comment (tomaszszymanski129) [inline saleor/graphql/attribute/utils.py:521]: same

File: saleor/graphql/attribute/utils.py [hunk 4]
```diff
@@ -750,15 +784,43 @@ def _pre_save_file_value(
...
+            (AttributeValue.objects.create, record) for record in values_to_create
+        ]
+
+    @classmethod
+    def _bulk_create_pre_save_values(cls, pre_save_bulk):
+        results: Dict["Attribute", list[AttributeValue]] = defaultdict(list)
+
+        for action, attribute_data in pre_save_bulk.items():
+            for attribute, values in attribute_data.items():
+                if action == AttributeValue.objects.create:
```
Comment (korycins) [inline saleor/graphql/attribute/utils.py:810]: Is there any reason why we use the manager method as an action type? Maybe some kind of enum would be better for that?

**Review by korycins — COMMENTED** [review]
Can we add a benchmark test for at least bulk mutation reported in #13700 (I don't see any)
It would be great also to have the info (as a PR comment), on how the improvement PR looks vs the target branch.

File: saleor/attribute/models/base.py [hunk 5]
```diff
@@ -215,6 +216,76 @@ def get_translated_keys(self):
...
+        objects_to_be_updated = []
+        update_fields = set()
+        query_params = []
+
+        for obj in objects_data:
+            defaults = obj.pop("defaults")
+            query_params.append(models.Q(**obj))
+            obj["defaults"] = defaults
+
+        query = self.filter(reduce(lambda x, y: x | y, query_params))
```
Comment (IKarbowiak) [inline saleor/attribute/models/base.py:260]:
```suggestion
        query_params = models.Q()

        for obj in objects_data:
            defaults = obj.pop("defaults")
            query_params |= models.Q(**obj)
            obj["defaults"] = defaults

        query = self.filter(query_params)
```

File: saleor/attribute/models/base.py [hunk 6]
```diff
@@ -215,6 +215,86 @@ def get_translated_keys(self):
         return {"name": self.name}
 
 
+class AttributeValueManager(models.Manager):
```
Comment (aniav) [inline saleor/attribute/models/base.py:218]: This class could use some comments. I remember debugging another manager somewhere else and it was quite hard to follow along. E.g. here it's not clear upon first look what's happening with the nested `for` in `bulk_get_or_create` 🙌

**Review by korycins — APPROVED** [review]
Thanks a lot for a great diff summary! ❤️

**Review by IKarbowiak — APPROVED** [review]

**Review by tomaszszymanski129 — APPROVED** [review]
